Dynamic Dashboard: Reviews Card Reload Functionality

hafizrahman opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of: #12742
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR includes:
- initial logic for fetching reviews (and products), and displaying it on the UI.
- functionality for viewing all reviews

The main purpose of this PR is to get more eyes on the reviews and products fetching logic, and make sure the logic is sound, before going forward with more functionality. Because of that, some functionalities are missing and will be added in subsequent PR:
1. The card doesn't load notifications yet, which is needed to determine balloon color (read or not read). The loading logic should be similar to products loading, so should be relatively simple to add once the products loading is confirmed.
2. Filtering is not done yet. Still discussing in Slack what's the best way forward.
3. Tapping each review is not added yet.
4. Empty and error states are not added yet.

## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

1. Run app, use a test site with at least 1 review on it.
5. Enable the "Most recent reviews" card in the dashboard if you haven't,
6. wait and ensure it loads and shows three reviews, with the correct subject, review content, and star.
7. Try the "View all reviews" and ensure it opens the Reviews list screen,
8. Go back to Dashboard and ensure the card still shows 3 cards as before.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

<img src="https://github.com/woocommerce/woocommerce-ios/assets/266376/fc8609dc-bafd-4a31-ab39-092d5b748f2c" width="42%">

hafizrahman committed:
Add functionality to open all reviews from card.

hafizrahman committed:
Add initial reloadData function for Reviews card.

hafizrahman committed:
Add db loading and display the result.

hafizrahman committed:
Strip HTML from the review text.

hafizrahman committed:
Use the existing ReviewViewModel for data.

hafizrahman committed:
Add functionality to fetch products after getting review product IDs, and add it to the data.

hafizrahman committed:
Update visibility handling.

hafizrahman committed:
Merge branch 'trunk'

Conflicts:
	WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewModel.swift
	WooCommerce/Classes/ViewRelated/Dashboard/Reviews/ReviewsDashboardCard.swift

hafizrahman committed:
Make sure to only get the latest three reviews.

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hafizrahman commented on WooCommerce/Classes/ViewRelated/Dashboard/Reviews/ReviewsDashboardCard.swift:
```diff
@@ -38,20 +40,26 @@ struct ReviewsDashboardCard: View {
-            ForEach(Array(dummyData.enumerated()), id: \.element.reviewID) { index, review in
-                ReviewRow(for: review, isLastItem: index == dummyData.count-1)
+            if viewModel.data.isNotEmpty {
+                ForEach(Array(viewModel.data.enumerated()), id: \.element.review.reviewID) { index, reviewViewModel in
```
The use of Array and enumerated here is so that we can get both `index` and `reviewViewModel` in the closure.

ref: https://stackoverflow.com/a/62589546

hafizrahman commented on WooCommerce/Classes/ViewRelated/Dashboard/Reviews/ReviewsDashboardCardViewModel.swift:
```diff
@@ -19,30 +19,173 @@ final class ReviewsDashboardCardViewModel: ObservableObject {
+    /// Updates data
+    @MainActor
+    func updateReviewsResults() async {
+        let reviews = productReviewsResultsController.fetchedObjects.prefix(3)
+        let reviewProductIDs = reviews.map { $0.productID }
+
+        // Load products that matches the review product IDs
+        if reviewProductIDs.isNotEmpty {
+            await loadReviewProducts(for: reviewProductIDs)
+        }
+
+        data = reviews
+            .map { review in
+                let product = reviewProducts.first { $0.productID == review.productID }
+                // TODO: also fetch notification
+                return ReviewViewModel(review: review, product: product, notification: nil)
+            }
```
I'd like to get more opinion on this area.

The fetch result for reviews only includes data for product ID related to the reviews, while the UI requires the product name to be displayed. Thus, another call for `loadReviewProducts()` is needed to get the product name. 

I'm not 100% sure if this is the best way to do this.

hafizrahman committed:
shorten name.

hafizrahman committed:
Simplify siteID predicate.

hafizrahman committed:
Set prefix with constant.

hafizrahman committed:
Add shimmer/redact to review items too when sync is happening.

hafizrahman commented on WooCommerce/Classes/ViewRelated/Dashboard/Reviews/ReviewsDashboardCardViewModel.swift:
```diff
@@ -19,30 +19,177 @@ final class ReviewsDashboardCardViewModel: ObservableObject {
+
+    /// Updates data
+    @MainActor
+    func updateReviewsResults() async {
+        let reviews = productReviewsResultsController.fetchedObjects.prefix(Constants.numberOfItems)
+        let reviewProductIDs = reviews.map { $0.productID }
+
+        // Load products that matches the review product IDs
+        if reviewProductIDs.isNotEmpty {
+            await loadReviewProducts(for: reviewProductIDs)
+        }
+
+        data = reviews
+            .map { review in
+                let product = reviewProducts.first { $0.productID == review.productID }
+                // TODO: also fetch notification
+                return ReviewViewModel(review: review, product: product, notification: nil)
+            }
+    }
```
I'd like to get more opinion on this area.

The fetch result for reviews only includes data for product ID related to the reviews, while the UI requires the product name to be displayed. Thus, another call `loadReviewProducts()` is needed to get the product names.

I'm not 100% sure if this is the best way to do this. It feels to me that it's going through many steps, but seems inevitable if we want to use storage as single source of truth:

1. remote call for reviews
2. handle db update for reviews
    1. remote call for products
    2. handle db update for products
    3. compare local reviews and local products to build data

selanthiraiyan replied:
Thanks for explaining, Hafiz!

I think you can try fetching from storage before sending a remote call for products.

You can use the following predicate to fetch multiple products from storage.

```
NSPredicate(format: "siteID == %lld AND productID IN %@", siteID, arrayOfProductIDs)
```

What do you think?

hafizrahman replied:
I updated the logic in https://github.com/woocommerce/woocommerce-ios/pull/12752/commits/4ad2e89c4dcf4bd4c51f0ba8662446e96aad4ccd 

Now it will fetch products from storage first and only do remote call if they're not found.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewModel.swift:
```diff
@@ -163,6 +163,9 @@ final class DashboardViewModel: ObservableObject {
                     await self?.mostActiveCouponsViewModel.reloadData()
                 }
+                group.addTask { [weak self] in
+                    await self?.reviewsViewModel.reloadData()
```
Suggestion: In a different PR review, Huong suggested hiding this behind a feature flag to avoid unnecessary reloads.

hafizrahman replied:
This is already in `featureFlagService.isFeatureFlagEnabled(.dynamicDashboardM2)`.  or did you mean a different feature flag? Can you share a link to that suggestion?

selanthiraiyan replied:
Oh, my bad. I missed to notice the feature flag check.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Dashboard/Reviews/ReviewsDashboardCard.swift:
```diff
@@ -106,27 +119,31 @@ private extension ReviewsDashboardCard {
     }
 
-    func ReviewRow(for review: ProductReview, isLastItem: Bool) -> some View {
+    func ReviewRow(for viewModel: ReviewViewModel, isLastItem: Bool) -> some View {
```
```suggestion
    func reviewRow(for viewModel: ReviewViewModel, isLastItem: Bool) -> some View {
```

hafizrahman replied:
Ah, nice! Subconsciously used Jetpack Compose habit here with capitalized name 😅 

Fixed in https://github.com/woocommerce/woocommerce-ios/pull/12752/commits/4fa99a1b9039bb6d0fdffa3af80b2be9a8950bb1 as it requires a change elsewhere too.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Dashboard/Reviews/ReviewsDashboardCard.swift:
```diff
@@ -38,20 +40,31 @@ struct ReviewsDashboardCard: View {
-                .padding(.horizontal, Layout.padding)
+            if viewModel.data.isNotEmpty {
+                ForEach(Array(viewModel.data.enumerated()), id: \.element.review.reviewID) { index, reviewViewModel in
+                    ReviewRow(for: reviewViewModel, isLastItem: index == dummyData.count-1)
```
```suggestion
                    ReviewRow(for: reviewViewModel, isLastItem: index == viewModel.data.count - 1)
```

hafizrahman replied:
Thanks! I fixed this in https://github.com/woocommerce/woocommerce-ios/pull/12752/commits/34094ecff5fd92c7ef1e7563a18c9ca874385c48 as I noticed it also myself before reading your comment.

hafizrahman committed:
Delete dummy data

selanthiraiyan reviewed (commented):
I left a comment about reading from storage before sending a request to retrieve products. Please let me know your thoughts.

hafizrahman committed:
Update incorrect dummy data usage.

hafizrahman committed:
Update func name.

hafizrahman commented on WooCommerce/Classes/ViewRelated/Dashboard/Reviews/ReviewsDashboardCard.swift:
```diff
@@ -38,20 +27,31 @@ struct ReviewsDashboardCard: View {
         .background(Color(.listForeground(modal: false)))
         .clipShape(RoundedRectangle(cornerSize: Layout.cornerSize))
         .padding(.horizontal, Layout.padding)
+        LazyNavigationLink(destination: ReviewsView(siteID: viewModel.siteID), isActive: $showingAllReviews) {
+            EmptyView()
+        }
```
This works but I'll refactor this to be handled in viewcontroller like other cards, in the next PR.

hafizrahman committed:
Update logic to get related products locally first, then fetch if needed.

hafizrahman committed:
Fix viewModel not being @ObservedObject

hafizrahman committed:
Simplify ForEach since we don't need index, just check last item.

hafizrahman committed:
Update the way reviewText snippet is shown in the swiftUI way.

hafizrahman committed:
More color fix and updates.

hafizrahman commented on WooCommerce/Classes/ViewRelated/Dashboard/Reviews/ReviewsDashboardCard.swift:
```diff
@@ -141,26 +146,24 @@ private extension ReviewsDashboardCard {
-            return Text(Localization.pendingReview).foregroundColor(Color(uiColor: .wooOrange)) +
-                Text(" • " + text)
+            return Text(pendingText).foregroundColor(Color(uiColor: accentColor)) +
+            Text(divider + text).foregroundColor(Color(uiColor: textColor))
```
here I wanted to use .foregroundStyle but Xcode said it's only available on iOS 17 above 🤔 
![Screenshot 2024-05-17 at 14 17 24](https://github.com/woocommerce/woocommerce-ios/assets/266376/73e2370e-6fc8-4242-8455-50e4d86c08a1)

selanthiraiyan replied:
I haven't used `+` to combine two `Text` before. Looks like we should use `AttributedString` instead.

hafizrahman replied:
I followed the guide here:

https://www.hackingwithswift.com/quick-start/swiftui/how-to-combine-text-views-together

hafizrahman replied:
I ended up updating to use AttributedString anyway to avoid using the deprecated `.foregroundColor()`. Thanks!

https://github.com/woocommerce/woocommerce-ios/pull/12752/commits/ae19a09781a0e1e40f2a672d0a85c6792f50a3b5

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Dashboard/Reviews/ReviewsDashboardCardViewModel.swift:
```diff
@@ -19,30 +19,174 @@ final class ReviewsDashboardCardViewModel: ObservableObject {
+
+    /// ResultsController for Product
+    private lazy var productsResultsController: ResultsController<StorageProduct> = {
+        let predicate = NSPredicate(format: "siteID == %lld AND productID IN %@", siteID, reviewProductIDs)
```
If we mistakenly access `productsResultsController` before assigning product IDs to `reviewProductIDs`, this predicate will not work. I don't see us reassigning the predicate again, either. 

We can either make this a non-lazy property or assign the predicate again once `reviewProductIDs` is ready.

hafizrahman replied:
That's right. I updated this in the refactor commit https://github.com/woocommerce/woocommerce-ios/pull/12752/commits/62be5a92c9e3dc767c0824aa26aefc3c13e08e2b

Now, it simply creates the `productsResultsController` on init and `updateProductsResultsControllerPredicate()` is added and called when product IDs are available for it.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Dashboard/Reviews/ReviewsDashboardCardViewModel.swift:
```diff
@@ -19,30 +19,174 @@ final class ReviewsDashboardCardViewModel: ObservableObject {
     }
 
+    @Published private(set) var data: [ReviewViewModel] = []
+    private var reviewProductIDs: [Int64] = []
```
Suggestion: If I am not wrong, I think we can skip storing this in a property and manage to send this value across functions as a parameter. Removing this property will improve reliability. One less property to follow. What do you think?

hafizrahman replied:
I agree. I removed it in the refactor commit https://github.com/woocommerce/woocommerce-ios/pull/12752/commits/62be5a92c9e3dc767c0824aa26aefc3c13e08e2b

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Dashboard/Reviews/ReviewsDashboardCardViewModel.swift:
```diff
@@ -19,30 +19,174 @@ final class ReviewsDashboardCardViewModel: ObservableObject {
+        try productsResultsController.performFetch()
+        reviewProducts = productsResultsController.fetchedObjects
+
+        if reviewProducts.isEmpty {
```
❓ If only a few products are available in storage, will we end up not loading the remaining products from remote?

hafizrahman replied:
Good catch. I'm updating the check in 

https://github.com/woocommerce/woocommerce-ios/pull/12752/commits/62be5a92c9e3dc767c0824aa26aefc3c13e08e2b

selanthiraiyan reviewed (commented):
Thanks for the updates to read from storage, @hafizrahman! 👍 

I left a few comments/questions about the logic. Please let me know your thoughts.

hafizrahman committed:
Various refactor for product fetching.

1. reviewProducts is now a computed var to simplify fetchProducts()
2. productsResultsController now initialized first and updated later when productIDs exist with updateProductsResultsControllerPredicate()
3. reviewProductIDs is now simply passed as function parameters and not kept as class variable
4. Update logic for checking existing products before needing to fetch
5. Code organization with MARK for Storage-related code

hafizrahman committed:
Merge branch 'trunk'

Conflicts:
	WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewModel.swift

hafizrahman committed:
Update review text to use AttributedString.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Dashboard/Reviews/ReviewsDashboardCardViewModel.swift:
```diff
@@ -19,30 +19,185 @@ final class ReviewsDashboardCardViewModel: ObservableObject {
+    public let siteID: Int64
     public let filters: [ReviewsFilter] = [.all, .hold, .approved]
 
+    private var productsResultsController: ResultsController<StorageProduct>
```
```suggestion
    private let productsResultsController: ResultsController<StorageProduct>
```

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Dashboard/Reviews/ReviewsDashboardCardViewModel.swift:
```diff
@@ -19,30 +19,185 @@ final class ReviewsDashboardCardViewModel: ObservableObject {
+
+    func updateProductsResultsControllerPredicate(with productIDs: [Int64]) {
+        let predicates = NSCompoundPredicate(andPredicateWithSubpredicates: [sitePredicate(),
+                                                                            NSPredicate(format: "productID IN %@", productIDs)])
```
Super nit: Indentation fix needed.

hafizrahman replied:
I think this looks odd but perhaps correct because it's an array as the parameter.

Either using enter manually or Editor > Structure > Re-indent results the same like above:

https://github.com/woocommerce/woocommerce-ios/assets/266376/445f0afc-45a6-484d-b386-0714cc9874c9

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Dashboard/Reviews/ReviewsDashboardCardViewModel.swift:
```diff
@@ -19,30 +19,185 @@ final class ReviewsDashboardCardViewModel: ObservableObject {
+
+        self.productsResultsController = ResultsController<StorageProduct>(storageManager: storageManager,
+                                                                           matching: nil,
+                                                                           sortedBy: [])
```
nit: We could mention the `fetchLimit: Constants.numberOfItems` here as well.

hafizrahman replied:
Good thinking. I had assumed that when the predicate is updated and it's requesting up to 3 product IDs, it would always return up to 3 items anyway since they're distinct. But I agree it's safer to limit it here too. Added in https://github.com/woocommerce/woocommerce-ios/pull/12752/commits/37fc4823dc944ec8a9b41c3a3cf4e11032b97085

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Dashboard/Reviews/ReviewsDashboardCardViewModel.swift:
```diff
@@ -19,30 +19,185 @@ final class ReviewsDashboardCardViewModel: ObservableObject {
+            stores.dispatch(ProductAction.retrieveProducts(siteID: siteID,
+                                                           productIDs: reviewProductIDs
+                                                          ) { result in
+                continuation.resume(with: result)
```
Non blocking suggestion:
We could replace this with the following to skip returning stuff from this function.

```swift
switch result {
 case .success:
  continuation.resume()
 case .failure(let error):
  continuation.resume(throwing: error)
}
```

The same applies to `func loadReviews() async throws -> [ProductReview]` method as well.

hafizrahman replied:
Thanks, I added the updates for these functions on https://github.com/woocommerce/woocommerce-ios/pull/12752/commits/40bb44abeae6843fcef40bbe860838ddd7a93dc0

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Dashboard/Reviews/ReviewsDashboardCardViewModel.swift:
```diff
@@ -19,30 +19,185 @@ final class ReviewsDashboardCardViewModel: ObservableObject {
+            // Ignoring the result from remote as we're using storage as the single source of truth
+            _ = try await retrieveProducts(for: reviewProductIDs)
+        } catch {
+            syncingError = error
```
Suggestion: I believe that we can display the reviews even if retrieving products fails. Could we ignore the error and just display the reviews without product info? 

If we think about it further, we could even load the products silently in the background and update the UI. What do you think?

hafizrahman replied:
This is a neat suggestion and I can try adding it on the next PR that deals with loading states. I do notice that it can take a while waiting for products and notifications to be fetched, and showing partial result first can be more useful to users. 👍🏼

selanthiraiyan reviewed (approved):
Thanks for the updates, Hafiz!

This looks good to me. I left a few non-blocking suggestions. 🚀

hafizrahman committed:
Update WooCommerce/Classes/ViewRelated/Dashboard/Reviews/ReviewsDashboardCardViewModel.swift

let/var fix

Co-authored-by: Sharma Elanthiraiyan <sharma.elanthiraiyan@automattic.com>

hafizrahman committed:
Add fetch limit to the products results controller initializer

hafizrahman committed:
Refactor remote fetch functions that do not need to return results.
